lib/services/downloads_service_backend.dart

```diff
@@ -0,0 +1,1515 @@
+  }
+
+  Future<void> _remove(IsarTaskDataType type, String? id) async {
+    _isar.writeTxnSync(() {
+      if (id != null) {
+        _isar.isarTaskDatas.deleteSync(IsarTaskData.getHash(type, id));
+      } else {
+        _isar.isarTaskDatas.where().typeEqualTo(type).deleteAllSync();
+      }
+    });
+  }
+}
+
+/// A wrapper for storing various types of download related data in isar as JSON.
+/// Do not confuse the id of this type with the ids that the content types have.
+/// They will not match.
+@collection
+class IsarTaskData<T> {
+  IsarTaskData(this.id, this.type, this.jsonData, this.age);
+  final Id id;
+  String jsonData;
+  @Enumerated(EnumType.ordinal)
+  @Index()
+  final IsarTaskDataType<T> type;
+  // This allows prioritization and uniqueness checking by delete buffer
+  // It is also used as a retry counter by sync buffer.
+  final int age;
+
+  static int globalAge = 0;
+
+  IsarTaskData.build(String stringId, this.type, T data, {int? age})
+      : id = IsarTaskData.getHash(type, stringId),
+        jsonData = _toJson(data),
+        age = age ?? globalAge++;
+
+  static int getHash(IsarTaskDataType type, String id) {
+    return _fastHash(type.name + id);
+  }
+
+  @ignore
+  T get data => type.fromJson(jsonDecode(jsonData));
+  set data(T item) => jsonData = _toJson(item);
+
+  static String _toJson(dynamic item) {
+    switch (item) {
+      case int id:
+        return jsonEncode({"id": id});
+      case (int itemIsarId, bool required, String? viewId):
+        return jsonEncode({
+          "stubId": itemIsarId,
+          "required": required,
+          "view": viewId,
+        });
+      case _:
+        return jsonEncode((item as dynamic).toJson());
+    }
+  }
+
+  /// FNV-1a 64bit hash algorithm optimized for Dart Strings
+  /// Provided by Isar documentation
+  static int _fastHash(String string) {
+    var hash = 0xcbf29ce484222325;
+
+    var i = 0;
+    while (i < string.length) {
+      final codeUnit = string.codeUnitAt(i++);
+      hash ^= codeUnit >> 8;
+      hash *= 0x100000001b3;
+      hash ^= codeUnit & 0xFF;
+      hash *= 0x100000001b3;
+    }
+
+    return hash;
+  }
+
+  @override
+  bool operator ==(Object other) {
+    return other is IsarTaskData && other.id == id;
+  }
+
+  @override
+  @ignore
+  int get hashCode => id;
+}
+
+/// Type enum for IsarTaskData
+/// Enumerated by Isar, do not modify existing entries.
+enum IsarTaskDataType<T> {
+  pausedTask<Task>(Task.createFromJson),
+  taskRecord<TaskRecord>(TaskRecord.fromJson),
+  resumeData<ResumeData>(ResumeData.fromJson),
+  deleteNode<int>(_deleteFromJson),
+  syncNode<(int, bool, String?)>(_syncFromJson);
+
+  const IsarTaskDataType(this.fromJson);
+
+  static int _deleteFromJson(Map<String, dynamic> map) {
+    return map["id"];
+  }
+
+  static (int, bool, String?) _syncFromJson(Map<String, dynamic> map) {
+    return (map["stubId"], map["required"], map["view"]);
```

Comment:
Would it be possible to use an enum for these strings, or are they to diverse/dynamic?

Reply:
They're just names of json fields to manually implement some to/from JSON methods. These are the only 4 values.  If you want this to be cleaner I can make the sync node into a class and use @JsonSerializable.

Reply:
That would be nice :)